compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3482,21 +3532,49 @@ public Boolean visit(BParameterizedType t, BType s) {
         }
     };
 
-    private boolean checkUnionHasSameType(LinkedHashSet<BType> memberTypes, BType baseType) {
+    private boolean isNil(BType type) {
+        BType referredType = getReferredType(type);
+        if (referredType.getKind() == TypeKind.NIL) {
+            return true;
+        } else if (referredType.getKind() == TypeKind.FINITE) {
+            Set<BLangExpression>  valueSpace = ((BFiniteType) referredType).getValueSpace();
+            return valueSpace.size() == 1 && valueSpace.iterator().next().getBType().getKind() == TypeKind.NIL;
+        }
+        return false;
+    }
+
+    private boolean checkUnionHasSameType(BUnionType unionType, BType baseType) {
+        LinkedHashSet<BType> memberTypes = unionType.getMemberTypes();
         boolean isSameType = false;
         for (BType type : memberTypes) {
             type = getReferredType(type);
             if (type.tag == TypeTags.FINITE) {
                 for (BLangExpression expr : ((BFiniteType) type).getValueSpace()) {
-                    isSameType = isSameOrderedType(expr.getBType(), baseType);
+                    isSameType = isSameOrderedType(expr.getBType(), baseType) || isNil(type);
```

Comment:
`checkUnionHasSameType` function checks whether there is any ordered type where the static type of both operands (unionType and baseType) belong. If there such type then it will return `true` and otherwise it will return `false`.

According to the spec `If type T is ordered, then type T? Is also ordered.`
Both `unionType` and `baseType` are ordered types since they were checked in previous stage (both operands are checked separately).
Ex. Let take base -> T, type(BFiniteType) -> (). T? is ordered type where the static type of both operands belong.
I had added this `isNil()` check because `isSameOrderedType()` returns false for some valid scenarios (expr.getBType() -> nilType and baseType -> arrayType).

I fixed this `isSameOrderedType()` and remove `isNil()` check since we don't need to make it special case.